### ISSUE TITLE
Sunodo run for testnet

### ISSUE THREAD

**Issue by tuler** [sunodo/sunodo#159]
## 📚 Context

Sunodo developed a new deployment procedure that is composed by two separate commands: `sunodo deploy` and `sunodo controller`.

Sunodo `deploy` is used by the developer to deploy his application to a testnet/mainnet. Sunodo `controller` is used by a validator to receive deploy requests and launch nodes for those applications. Those two actors, developer and validator, and separated by a web3 cloud, in a decentralized fashion.

We use IPFS for Cartesi machine distribution, and the base layer (Ethereum) as the deploy platform.

In addition to that Sunodo introduces a financial layer to provide incentives for the validator to launch nodes. This is implemented using a ERC-20 token chosen by the validator.

For the scenario where the developer and the validator are a single person this scenario could be much simplified.
The Cartesi machine is already available and we don't need IPFS.
There is no need for a financial incentive, as they are the same person.

## ✔️ Solution

Extend the `sunodo run` command to execute nodes for testnet, where the machine is located in the local disk, and there is no financial mechanism.

We have to modify the current behavior to exclude the `anvil` node from the stack, and require a web3 gateway configuration.

We should support this only for testnet, for a few reasons:

- the database and redis of the environment can still be there, but they are ephemeral
- we need a wallet. We can still support MNEMONIC even though it's very unsafe
- we can use a consensus that is open to anyone to submit claims, so we don't need to deploy an Authority

## 📈 Subtasks

- [ ] implement a test consensus open to anyone to submit claims
- [ ] extend the CLI compose file to allow testnets
- [ ] extend `sunodo run` with testnet options

## 🎯 Definition of Done

- [ ] `sunodo run --rpc-url <testnet_url> --mnemonic-passphrase ...` working for testnet

**Comment by tuler** [comment]
I wanted to understand better why people want this, the use cases, and what are their expectations.

I cat think of the following reasons:

1) "I want to have an environment which behaves closer to how a testnet behaves".

Anvil has options to control its behavior to mimic better a testnet network. Block time, mining mode, gas price can all be tuned to make it work more like a testnet.

2) "I want to deploy the application to a testnet so other people can use it".

This is a fair reason. But I feel that people often forgets that deploying a dapp to a testnet is just the first step. There must be a node running for that dapp, and its endpoints (graphql and inspect) must be publicly available. So running on the developer's computer may not be the best solution, even for simple demonstrations scenarios.

**Comment by miltonjonat** [comment]
This is an interesting discussion!

1. Fine, although in this case you would either want to replicate those Anvil mimicking options in `sunodo run`, or else have a `--no-anvil` mode because you will spin up your own Anvil. For the latter case, maybe just having `--rpc-url` would be the way to specify that you don't want sunodo to spin up an Anvil internally, which is pretty much what this issue is proposing anyway. On top of all that, I also think that "mimicking better" is a lot different than the actual thing.

2. I mostly agree. In sunodo, "deploying to a testnet so other people can use it" will be achieved by executing `sunodo deploy`. But we must also bear in mind that "demonstration scenarios" may be a common activity, and I'm not sure we want sunodo cloud to keep spending resources to instantiate these test nodes all the time. Do we?

I can also think of a different reason:

3. "I want to deploy the application to a testnet where other contracts are already deployed and/or other state is defined, with which I want to interact"

This could mean existing asset contracts (ERC-20, ERC-721) or other contracts for which vouchers will be generated (e.g., Uniswap, whatever). And when we have the dehashing device, I can see this being even more relevant.
Of course, meeting this demand can be achieved by running a local Anvil with `--fork-url`, and pointing sunodo run to it. So I guess in the end it falls within case (1), in which you would need `--rpc-url` to specify where your local or remote Anvil (or whatever else) with the desired fork is running - or else you would need to also have a `--fork-url` option in `sunodo run`.

**In conclusion**: I think specifying `--rpc-url` is a good thing to have, and let the user configure whatever he wants in that URL. In that case, I don't see why we would want to restrict this to localhost only, and why we would prevent him from connecting directly to a testnet. I think the sunodo semantics of `sunodo run` vs `sunodo deploy` will make it clear that, when you decide to run specifying a remote RPC node as your base layer, you're **NOT** deploying the DApp to a testnet or anything

**Comment by tuler** [comment]
One thing to consider: `sunodo run` does a few things:

1) run a local anvil with rollups deployed
2) deploy an authority
3) deploy the dapp
4) run the node

What would the `sunodo run --rpc <anything>` do?
Not 1.
Not 2?
Not 3?
Just 4?

It's still confusing.

**Comment by tuler** [comment]
The alternative would be:

1) run you own anvil
2) deploy rollups yourself
3) deploy authority yourself
4) run `sunodo deploy`
5) use the deploy information to `sunodo run --rpc <url>`

Sunodo deploy would have two options: an unincentivized mode and an incentivized mode.

**Comment by miltonjonat** [comment]
I'm not sure I followed. So there *would* be a --rpc option, but it would do only 4 then?

**Comment by miltonjonat** [comment]
Fair enough, I guess that covers the use cases then?

**Comment by tuler** [comment]
This has been superseded by https://github.com/sunodo/sunodo/issues/383
